added state and multilevel prolong + tests.

alsgregory opened the pull request:
Initial code commit. Including:

- travis, make and setup
- __init__.py and the source code for the state object
- methods for recursive (multiple level) prolonging and injecting of functions in a hierarchy
- tests for these
- demos

miklos1 commented on .travis.yml:
```diff
@@ -0,0 +1,42 @@
+script:
+  - make lint
+  - make test
+
```
please...

alsgregory replied:
Noted. Isn't it ironic that the script checking linting errors isn't checked for this itself...

alsgregory replied:
done on new commit

wence- commented on .travis.yml:
```diff
@@ -0,0 +1,39 @@
+notifications:
+  slack:
+    secure: cuB9waMrZiRnJjJLq68NTSg1nB+ypbCTy4cphu0VjfuDYhmQc4MgnkvmzLntsE9X9Oj8kSP51rRjD+IJJ8qT1GTMi/5XAqKNzncONEdx5T0Tict0xxAaY8OcumOV68tisQ1jEeAVA65zcvPoyWOVv+ha4MdHLx8+J85h/zOFMRALf2tJ8tUnC2iHidWCYsQS8vQEkKp1eDWJRdcJY+OrmeV5tlgWXP56b6pRdnhoWwOkwtRols6yoPw1j6uNewb3N2uvJIIw5g7ji209nXFM+/HFGLLqNlWUpqickbaZEG/m5VqeuxeaZTl6jlW2w2+B5zyMPKR8NHeSQ8+c6Si08LbgQDGkLh1xHNQOTxqTEOHUOsLgBm7Y+NX1QDfIfclvY+d ...
```
This is the wrong secure string.  Remove the notifications section and I will add it in a separate pull request.

wence- commented on .travis.yml:
```diff
@@ -0,0 +1,39 @@
+  - curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install
+  - python ./firedrake-install --disable-ssh --minimal-petsc
+  - . ./firedrake/bin/activate
+  - python setup.py install
```
I would run `pip install .` instead.

alsgregory commented:
I've changed travis file for those corrections. I've also just changed the state.py slightly to remove a setting where one could carry lists of Functions. This over complicated things for now. Is it ok to merge after this?

alsgregory commented:
Major update to initial commit. After conversations about prolonging and injecting over recursive levels (this functionality has been pushed to a branch in firedrakeproject/firedrake). 

Removed:
- prolonging and injecting over multiple levels is removed and incorporated into multigrid

Added:
- Generalised hierarchies are able to build FS and Mesh hierarchies that are refined by a factor greater than two.
- tests and demos

Altered:
- no more dependence on FunctionHierarchy, which will be removed from multigrid
- state.py removing unnecessary error checks

alsgregory commented:
Can I merge?
